p2p/conn/connection_test.go

```diff
@@ -112,6 +115,56 @@ func TestMConnectionSend(t *testing.T) {
 	assert.False(t, mconn.Send(0x05, []byte("Absorbing Man")), "Send should return false because channel is unknown")
 }
 
+func TestMConnectionSendRate(t *testing.T) {
+	server, client := NetPipe()
+	defer server.Close()
+	defer client.Close()
+
+	clientConn := createTestMConnection(client)
+	err := clientConn.Start()
+	require.Nil(t, err)
+	defer clientConn.Stop() //nolint:errcheck // ignore for tests
+
+	// prepare a message to send from client to the server
+	msg := bytes.Repeat([]byte{1}, 1000*1024)
+
+	// send the message and check if it was sent successfully
+	done := clientConn.Send(0x01, msg)
+	assert.True(t, done)
+
+	// read the message from the server
+	_, err = server.Read(make([]byte, len(msg)))
+	if err != nil {
+		t.Error(err)
+	}
+
+	// check if the peak send rate is within the expected range
+	peakSendRate := clientConn.Status().SendMonitor.PeakRate
+	// the peak send rate should be less than or equal to the max send rate
+	// the max send rate is calculated based on the configured SendRate and other configs
+	maxSendRate := clientConn.maxSendRate()
+	assert.True(t, peakSendRate <= clientConn.maxSendRate(), fmt.Sprintf("peakSendRate %d > maxSendRate %d", peakSendRate, maxSendRate))
+}
+
+// maxSendRate returns the maximum send rate in bytes per second based on the MConnection's SendRate and other configs. It is used to calculate the highest expected value for the peak send rate.
+// The returned value is slightly higher than the configured SendRate.
+func (c *MConnection) maxSendRate() int64 {
+	sampleRate := 100 * time.Millisecond
```

Comment:
is this a constant somehwere and if so could we reference it?

Reply:
It's not a constant; it can be configured for each connection when it's created. However, the field containing this value was originally private. As a workaround, I duplicated the value in a hardcoded format within this function. Nevertheless, I believe it would be beneficial to include a utility function to make that field accessible and use it instead. I have added this feature in the following commits https://github.com/celestiaorg/celestia-core/pull/1112/commits/7616d8ec022a390fdaf7423e5ecda58bc9a53554 and https://github.com/celestiaorg/celestia-core/pull/1112/commits/e9e374ab72843bed26ec7592f90b5ba7b2dc127d.

Reply:
Alright, the recent modifications I made are considered breaking changes and would necessitate adjustments to several other test files. Therefore, I've decided to roll back those changes for now https://github.com/celestiaorg/celestia-core/pull/1112/commits/f410d352efeb4181d4d37b9ed99f280dd7903322. I will create a separate follow-up PR to introduce these changes independently in order to maintain cleaner and more focused PR scopes.

Reply:
In this [commit](https://github.com/celestiaorg/celestia-core/pull/1112/commits/d4c3074aa45605e42bd060763195af025700c4b8), I have included a comment to clarify the source of the sample rate.

For reference, you can also view the relevant code [here](https://github.com/celestiaorg/celestia-core/blob/e4f6117c26750e730d7a83b543cce28698c918b4/p2p/conn/connection.go#L182). 
The first argument passed to `flow.New(0, 0)` represents the sample rate, which subsequently translates to `100 ms` within the `New` method.

Reply:
Tracking issue https://github.com/celestiaorg/celestia-core/issues/1114